Purchases/SwiftSources/IntroEligibilityCalculator.swift

```diff
@@ -0,0 +1,94 @@
+//
+//  IntroEligibilityCalculator.swift
+//  Purchases
+//
+//  Created by Andrés Boedo on 7/14/20.
+//  Copyright © 2020 Purchases. All rights reserved.
+//
+
+import Foundation
+import StoreKit
+
+@objc(RCIntroEligibilityCalculator) class IntroEligibilityCalculator: NSObject {
+    private let productsManager: ProductsManager
+    private let receiptParser: ReceiptParser
+    
+    override init() {
+        self.productsManager = ProductsManager()
+        self.receiptParser = ReceiptParser()
+    }
+    
+    internal init(productsManager: ProductsManager,
+                  receiptParser: ReceiptParser) {
+        self.productsManager = productsManager
+        self.receiptParser = receiptParser
+    }
+    
+    @available(iOS 12.0, macOS 10.14, macCatalyst 13.0, tvOS 12.0, watchOS 6.2, *)
+    @objc func checkTrialOrIntroductoryPriceEligibility(with receiptData: Data,
+                                                        productIdentifiers candidateProductIdentifiers: Set<String>,
+                                                        completion: @escaping ([String: NSNumber], Error?) -> Void) {
+        guard candidateProductIdentifiers.count > 0 else {
+            completion([:], nil)
+            return
+        }
+        
+        var result: [String: NSNumber] = candidateProductIdentifiers.reduce(into: [:]) { resultDict, productId in
+            resultDict[productId] = IntroEligibilityStatus.unknown.toNSNumber()
+        }
+        do {
+            let receipt = try receiptParser.parse(from: receiptData)
+            let purchasedProductIdsWithIntroOffers = receipt.purchasedIntroOfferOrFreeTrialProductIdentifiers()
+            
+            let allProductIdentifiers = candidateProductIdentifiers.union(purchasedProductIdsWithIntroOffers)
```

Comment:
join all product ids so that we only issue one products request, we separate them again in the completion block